src/Module/Contact/Profile.php

```diff
@@ -0,0 +1,464 @@
+			throw new HTTPException\NotFoundException(DI::l10n()->t('Contact not found.'));
+		}
+
+		$localRelationship = DI::localRelationship()->getForUserContact(local_user(), $contact['id']);
+
+		if ($localRelationship->rel === Contact::SELF) {
+			DI::baseUrl()->redirect('profile/' . $contact['nick'] . '/profile');
+		}
+
+		if (isset($parameters['action'])) {
+			self::checkFormSecurityTokenRedirectOnError('contact/' . $contact['id'], 'contact_action', 't');
+
+			$cmd = $parameters['action'];
+			if ($cmd === 'update' && $localRelationship->rel !== Contact::NOTHING) {
+				Module\Contact::updateContactFromPoll($contact['id']);
+			}
+
+			if ($cmd === 'updateprofile' && $localRelationship->rel !== Contact::NOTHING) {
+				self::updateContactFromProbe($contact['id']);
+			}
+
+			if ($cmd === 'block') {
+				if ($localRelationship->blocked) {
+					// @TODO Backward compatibility, replace with $localRelationship->unblock()
+					Contact\User::setBlocked($contact['id'], local_user(), false);
+
+					$message = DI::l10n()->t('Contact has been unblocked');
+				} else {
+					// @TODO Backward compatibility, replace with $localRelationship->block()
+					Contact\User::setBlocked($contact['id'], local_user(), true);
+					$message = DI::l10n()->t('Contact has been blocked');
+				}
+
+				// @TODO: add DI::localRelationship()->save($localRelationship);
+				info($message);
+			}
+
+			if ($cmd === 'ignore') {
+				if ($localRelationship->ignored) {
+					// @TODO Backward compatibility, replace with $localRelationship->unblock()
+					Contact\User::setIgnored($contact['id'], local_user(), false);
+
+					$message = DI::l10n()->t('Contact has been unignored');
+				} else {
+					// @TODO Backward compatibility, replace with $localRelationship->block()
+					Contact\User::setIgnored($contact['id'], local_user(), true);
+					$message = DI::l10n()->t('Contact has been ignored');
+				}
+
+				// @TODO: add DI::localRelationship()->save($localRelationship);
+				info($message);
+			}
+
+			DI::baseUrl()->redirect('contact/' . $contact['id']);
+		}
+
+		$vcard_widget  = Widget\VCard::getHTML($contact);
+		$groups_widget = '';
+
+		if (!in_array($localRelationship->rel, [Contact::NOTHING, Contact::SELF])) {
+			$groups_widget = Group::sidebarWidget('contact', 'group', 'full', 'everyone', $contact['id']);
+		}
+
+		DI::page()['aside'] .= $vcard_widget . $groups_widget;
+
+		$o = '';
+		Nav::setSelected('contact');
+
+		$_SESSION['return_path'] = DI::args()->getQueryString();
+
+		DI::page()['htmlhead'] .= Renderer::replaceMacros(Renderer::getMarkupTemplate('contact_head.tpl'), [
+			'$baseurl' => DI::baseUrl()->get(true),
+		]);
+
+		$contact['blocked']  = Contact\User::isBlocked($contact['id'], local_user());
+		$contact['readonly'] = Contact\User::isIgnored($contact['id'], local_user());
+
+		switch ($localRelationship->rel) {
+			case Contact::FRIEND:   $relation_text = DI::l10n()->t('You are mutual friends with %s', $contact['name']); break;
```

Comment:
I notice and like the distinction between the "core" `$localRelationship->rel` content and it's localized presentation text inside the module :-)